core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -91,9 +89,9 @@ class TransactionStateManager(brokerId: Int,
   // we will get the lock when actually trying to transit the transaction metadata to abort later.
   def transactionsToExpire(): Iterable[TransactionalIdAndProducerIdEpoch] = {
     val now = time.milliseconds()
-    transactionMetadataCache.flatMap { case (_, entry) =>
+    transactionMetadataCache.flatMap { case (txnPartition, entry) =>
```

Comment:
EDIT: actually we do need the read lock not for correctness but to avoid ConcurrentModificationException.